### DISCUSSION TITLE
SurfaceView Preview and Stream is Upside-down

### DISCUSSION THREAD

**Discussion by Taiki1123** [pedroSG94/RootEncoder#1260] [Q&A] [answered]
Thank you for your module, It is very helpful.

I'm making an app that livestreams a camera with UVC input.
By setting CameraID to 2 in RtmpCamera1, I was able to use the UVC input camera.
However, the image is flipped upside down on the screen previewed on SurfaceView and the screen distributed to Youtube etc.
The rotation is set to 0, and setting it to any other value does not change the upside-down problem.
How can I display it without flipping?
Thank you.

**Comment by pedroSG94** [comment]
Hello,
If you want to flip preview/stream you need to use openglview.

1- Replace SurfaceView to openglview
2 - flip the preview/stream:
You have 4 methods to flip preview/stream:
```
rtmpCamera1.getGlInterface().setIsPreviewHorizontalFlip(true);
rtmpCamera1.getGlInterface().setIsPreviewVerticalFlip(true);
rtmpCamera1.getGlInterface().setIsStreamHorizontalFlip(true);
rtmpCamera1.getGlInterface().setIsStreamVerticalFlip(true);
```
I think it is autoexplicative but:

Only affect to preview:
setIsPreviewHorizontalFlip
setIsPreviewVerticalFlip

Only affect to stream:
setIsStreamHorizontalFlip
setIsStreamVerticalFlip

**Reply by Taiki1123** [reply]
Thank you for your answer.
I have follow-up questions.

Is it normal for SurfaceView to have different top and bottom for preview and stream? This is my first problem.
Supplementally, the top and bottom were aligned in the method using OpenGL.

i modify onPreviewFrame of "Camera1ApiManager.class" to export bitmap data.
However, onPreviewFrame doesn't work when using OpenGL.
So I'd like to use a SurfaceView that fixed the problem, or somehow use OpenGL to export bitmaps.

I would like some ideas for beginners.

**Comment by pedroSG94** [comment, marked as answer]
Hello,

> Is it normal for SurfaceView to have different top and bottom for preview and stream? This is my first problem.
> Supplementally, the top and bottom were aligned in the method using OpenGL.

Yes, using camera1 with SurfaceView or TextureView the preview and buffer data is not allways aligned because the buffer provided and the preview came from differents ways. Using camera2 or OpenGlView the preview and the stream is the same because both share the same source.
I only had a device with this problem in my hand (Nexus 6P) so it is not the normal but it is possible. You can fix it rotating the buffer byte[] but since it only happens in few devices use camera2 or OpenGlView is better and the performance is also better.

> i modify onPreviewFrame of "Camera1ApiManager.class" to export bitmap data.
> However, onPreviewFrame doesn't work when using OpenGL.
> So I'd like to use a SurfaceView that fixed the problem, or somehow use OpenGL to export bitmaps.

Using OpenGlView you can use this method to capture bitmaps:
```java
rtmpCamera1.getGlInterface().takePhoto(new TakePhotoCallback() {
          @Override
          public void onTakePhoto(Bitmap bitmap) {
            
          }
        });
```

**Reply by Taiki1123** [reply]
Thank you for your reply.
Now I understand why upside down happens. thank you.
However, onTakePhoto is not executed even though it is implemented.
After setting onTakePhoto to TakePhotoCallback, does this always run after startPreview?
I apologize for the elementary question, but thank you in advance.

**Comment by pedroSG94** [comment]
Hello,

The code example posted above must be called after startPreview and each time you want to get a bitmap. For example, when you click a button you can paste that code in the event.

**Reply by Taiki1123** [reply]
I could understand your tips and solve this problem.
thank you very much to your support.
